core/snmp/profile-mapper/src/main/java/org/opennms/core/snmp/profile/mapper/impl/SnmpProfileMapperImpl.java

```diff
@@ -0,0 +1,176 @@
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.core.snmp.profile.mapper.impl;
+
+import java.net.InetAddress;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.opennms.netmgt.config.api.SnmpAgentConfigFactory;
+import org.opennms.netmgt.config.snmp.SnmpProfile;
+import org.opennms.netmgt.filter.api.FilterDao;
+import org.opennms.netmgt.filter.api.FilterParseException;
+import org.opennms.netmgt.snmp.SnmpAgentConfig;
+import org.opennms.netmgt.snmp.SnmpProfileMapper;
+import org.opennms.netmgt.snmp.SnmpValue;
+import org.opennms.netmgt.snmp.proxy.LocationAwareSnmpClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import com.google.common.base.Strings;
+
+public class SnmpProfileMapperImpl implements SnmpProfileMapper {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SnmpProfileMapperImpl.class);
+
+    @Autowired
+    private SnmpAgentConfigFactory agentConfigFactory;
+
+    @Autowired
+    private FilterDao filterDao;
+
+    @Autowired
+    private LocationAwareSnmpClient locationAwareSnmpClient;
+
+    public SnmpProfileMapperImpl(FilterDao filterDao, SnmpAgentConfigFactory agentConfigFactory, LocationAwareSnmpClient locationAwareSnmpClient) {
+        this.agentConfigFactory = Objects.requireNonNull(agentConfigFactory);
+        this.filterDao = Objects.requireNonNull(filterDao);
+        this.locationAwareSnmpClient = Objects.requireNonNull(locationAwareSnmpClient);
+    }
+
+    public SnmpProfileMapperImpl() {
+    }
+
+    private static final String SYS_OBJECTID_INSTANCE = ".1.3.6.1.2.1.1.2.0";
+
+    @Override
+    public Optional<SnmpAgentConfig> getAgentConfigFromProfiles(InetAddress inetAddress, String location, String oid) {
+
+        List<SnmpProfile> snmpProfiles = agentConfigFactory.getProfiles();
+        // Get matching profiles for this IpAddress.
+        List<SnmpProfile> matchedProfiles = snmpProfiles.stream()
+                .filter(snmpProfile -> isFilterExpressionValid(inetAddress, snmpProfile.getFilterExpression()))
+                .collect(Collectors.toList());
+
+        for (SnmpProfile snmpProfile : matchedProfiles) {
+
+            Optional<SnmpAgentConfig> config = fitProfile(snmpProfile, inetAddress, location, oid);
```

Comment:
Profiles should be fitted in parallel instead of serially (take advantage of the fact the SNMP lookups are already async.). When all have returned, we can iterate through the results using the same order as which the profiles are defined in the configuration file and take the first that matched.